### ISSUE TITLE
Process parsing breaks on Linux when names include spaces

### ISSUE THREAD

**Issue by newk5** [oshi/oshi#442]
I noticed that I can sort the running processes by memory with:

`OperatingSystem.ProcessSort.MEMORY`

I'm under the impression this sorting is done by shared memory (SHR) correct me if I'm wrong about this, but is it possible to sort by Resident Set Size (RSS) instead ? 

Thanks!

**Comment by dbwiddis** [comment]
The `MEMORY` sort is done by RSS.  See the [AbstractOperatingSystem](https://github.com/oshi/oshi/blob/master/oshi-core/src/main/java/oshi/software/common/AbstractOperatingSystem.java#L48) class.

**Comment by newk5** [comment]
Oh, I see. Hmm, maybe I'm doing something wrong. I'm currently running ElementaryOS 0.4.1 Loki and I'm using:` OSProcess topMem = os.getProcesses(1, OperatingSystem.ProcessSort.MEMORY)[0]` to to get the top process using the most memory but for some reason it doesn't seem to pick up the right process. Using `htop` if I sort by RES I can see it shows Netbeans at the top. I also  tried using a task manager like utility program and it too shows NetBeans at the top:
![screenshot from 2018-01-29 00 58 00](https://user-images.githubusercontent.com/32560510/35489538-bb0aba3e-048f-11e8-9c75-ce79e583ccdc.png)

But when using oshi, I always get firefox as the top process. Is there something I'm doing wrong?

**Comment by dbwiddis** [comment]
On Linux, OSHI pulls the RSS value from the 24th field of `/proc/pid/stat`.  (The value is in pages.)  Can you check this value for the NetBeans and FireFox processes and see what it says, compared to `htop`?

**Comment by dbwiddis** [comment]
So OSHI should report RSS of 4096 x 408420 = 1.53 GiB for pid 4086 (java/netbeans), and 4096 x 100352 = 392 MiB for pid 2913 (firefox).  

Is it returning the correct memory size for firefox?   Does it even see the netbeans process?

**Comment by newk5** [comment]
I've tried this:

```
List<OSProcess> procs = Arrays.asList(os.getProcesses(10, ProcessSort.MEMORY));

        System.out.println("   PID  %CPU %MEM       VSZ       RSS Name");
        for (int i = 0; i < procs.size() && i < 20; i++) {
            OSProcess p = procs.get(i);
            System.out.format(" %5d %5.1f %4.1f %9s %9s %s%n",
                    p.getProcessID(),
                    100d * (p.getKernelTime() + p.getUserTime()) / p.getUpTime(),
                    100d * p.getResidentSetSize() / memory.getTotal(),
                    FormatUtil.formatBytes(p.getVirtualSize()),
                    FormatUtil.formatBytes(p.getResidentSetSize()), p.getName());
        }
```
And got this
```
  PID  %CPU %MEM       VSZ       RSS Name
  2969   2.8 56843.9   2.4 KiB   8.6 TiB Web
  6211   0.4 54019.3 740.4 KiB   8.2 TiB Web
 27744   0.2 51885.6   2.2 MiB   7.9 TiB Web
 27827   0.1 50951.4   2.2 MiB   7.7 TiB Web
  4086   4.7 10.3   6.8 GiB   1.6 GiB java
  2913   2.6  2.5   2.5 GiB 390.8 MiB firefox
  4158   0.5  1.6   4.0 GiB 249.2 MiB pantheon-mail
  4589   0.8  1.2   1.9 GiB 194.5 MiB Discord
 17000   0.1  1.1   1.2 GiB 180.1 MiB code
  1042   4.8  1.1 555.5 MiB 178.4 MiB Xorg
```
I believe those 4 top processes are all from firefox, I think I'm getting process 2969 when I do:
`OSProcess topMem = os.getProcesses(1, OperatingSystem.ProcessSort.MEMORY)[0]`
It seems like on that list the netbeans process (4086) is correctly above the firefox process (2913).

On htop the 2969 process shows like this:
![screenshot from 2018-01-29 06 38 03](https://user-images.githubusercontent.com/32560510/35496685-00ce6fb4-04bf-11e8-87c6-209e5eeae107.png)

Possibly a child process from firefox? Although on the task manager utility I'm using these child processes don't show up. Is there a way for me to exclude them from my oshi ProcessSort ?

**Comment by dbwiddis** [comment]
Interesting those top processes have an RSS far in excess of your 16GB available memory!  

You can manually sort the process list yourself (see the OSHI code linked above for an example) and exclude them; the question is how/why to exclude them.  OSHI has a parent PID field if they're child processes, but I'm not sure that's what's going on here.  

Something else is obviously weird going on to produce such high RSS values for those processes.  Can you do a `/proc/pid/stat` printout for them?  I am suspecting they somehow have a different format/number of fields and aren't matching (thus OSHI is picking a different field to calculate RSS).

**Comment by YoshiEnVerde** [comment]
Ok, so the top 4 processes seem to have memory values in excess of 3 to 6 orders of magnitude of what they should probably have.

That obviously means we're probably parsing the wrong value there.

Believing that, I checked over the code, and I found an issue (at least in the latest values posted by @newk5 ):
We parse the **proc/PID/stat** file by splitting whitespaces. However, the top (wrongly parsed) 4 processes are named `Web Content`.
This causes us to get two elements from the process name, instead of one: `(Web` and `Content)`

This means we're parsing the **_VSZ_** as the **_RSS_**, and the **_Start Time_** a the **_VSZ_**.
That leaves us with a timestamp in clock ticks, formatted as a value in bytes (ST>VSZ) and a mem size in bytes formatted as a mem size in pages (VSZ>RSS).

If I reformat the wrong value presented for RSS as VSZ, I get:
```
  PID       VSZ
  2969     2.2 GiB
  6211     2.1 TiB
  27744   2.0 TiB
  27827   1.9 TiB
```
That seems a bit more realistic for the 16GiB RAM we were working over. It even matches the _htop_ value

**Comment by YoshiEnVerde** [comment]
Of course, this still doesn't address the original issue of how the NetBeans and Firefox processes were showing unsorted, but it's a nice bug to catch.

I think we should do a more specific split patter for our processes. Specially since [http://man7.org/linux/man-pages/man5/proc.5.html](url) shows us the exact format we're expecting from the files.

At the very least, since all values are numeric, with only the proc name being a string, we could use something like:

1. Split the PID away: `\\s+\\(`
2. Split the Name away: `\\)\\s+`
3. Split the rest away: `\\s+`

instead of

- Split the everything away: `\\s+`

**Comment by dbwiddis** [comment]
I think you nailed it, Yoshi.  I was on my way there asking for the `proc/pid/stat` but knowing now the "Web Content" name, that's clearly the culprit.

I believe the process name is within parenthesis, so we can still parse everything out; we just need to do the parenthesis first.    I think you're on track with the splitting (I would avoid regexp for performance).  Do you want to take a crack at the PR?

**Comment by YoshiEnVerde** [comment]
Sure, I'll do the PR.

BTW, String implementations usually build a `Pattern` file for each call to `split()`. That was why we did the performance update refactoring our whitespace splits to a pattern in `ParseUtil`, so I'll just add two more patterns in there, then use them to do the three chained splits.

**Comment by newk5** [comment]
Ah nice to see the issue is identified, not sure if it's still neeeded but here's a `proc/pid/stat` of one of the Firefox Web content processes anyway:
`4235 (Web Content) S 4040 3749 3702 34817 3749 4194304 121952 0 3 0 2252 290 0 0 20 0 41 0 111337 2119409664 62149 18446744073709551615 94290627489792 94290627688956 140729235825792 0 0 0 0 4098 1107297528 0 0 0 17 6 0 0 79755 0 0 94290629790128 94290629791856 94290650783744 140729235831653 140729235833057 140729235833057 140729235836895 0`

pid is diferent because I rebooted.

I think I might be able to clarify this:
> Of course, this still doesn't address the original issue of how the NetBeans and Firefox processes were showing unsorted

When on my first post I said I was getting firefox as the top process instead of Netbeans I was actually refering to one of those firefox child web content processes, not the parent firefox process itself. Sorry about that maybe I should've been clearer :)

**Comment by dbwiddis** [comment]
Thanks for the clarification, newk5.  I had assumed as much by this point, but good to hear it.  Nice catch!

**Comment by dbwiddis** [comment]
Also your string field would be a perfect test case for the junit test that Yoshi will be writing. ;)

**Comment by YoshiEnVerde** [comment]
PR for this issue created at #443

**Comment by dbwiddis** [comment]
Fixed in #443
